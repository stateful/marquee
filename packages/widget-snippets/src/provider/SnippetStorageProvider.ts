import path from 'path'
import vscode from 'vscode'
import { EventEmitter } from 'events'
import { Logger, ChildLogger } from '@vscode-marquee/utils/extension'

import Snippet from '../models/Snippet'
import { STATE_KEY } from '../constants'
import type { State } from '../types'

export default class SnippetStorageProvider extends EventEmitter implements vscode.FileSystemProvider {
  static scheme = 'snippet'

  #logger: ChildLogger
  private _emitter = new vscode.EventEmitter<vscode.FileChangeEvent[]>()
  readonly onDidChangeFile: vscode.Event<vscode.FileChangeEvent[]> = this._emitter.event

  constructor (
    private _context: vscode.ExtensionContext,
    private _workspaceId?: string
  ) {
    super()
    this.#logger = Logger.getChildLogger('SnippetStorageProvider')
  }

  watch (): vscode.Disposable {
    return new vscode.Disposable(() => { })
  }

  stat (uri: vscode.Uri): Snippet {
    if (uri.path.slice(1) === 'New Clipboard Item') {
      return new Snippet(this._workspaceId)
    }

    const state = this._context.globalState.get<State>(STATE_KEY)
    let snippet = state?.snippets.find((snippet) => (snippet.storagePath || snippet.path) === uri.path)

    /**
     * in Marquee v2 and earlier `path` was used to recognise the source of the snippet.
     * This has changed in v3 where `path` represents the virtual path and `origin` the
     * source of the snippet. To allow finding old snippets (e.g. imported in v3) we do
     * this extra check.
     * In v4 the storage path in now written into `storagePath` property.
     */
    if (!snippet) {
      const [id, path] = uri.path.split('/').filter(Boolean)
      snippet = state?.snippets.find((snippet) => snippet.id === id && (snippet.storagePath || snippet.path) === path)
    }

    if (!snippet) {
      const message = `Couldn't find snippet at ${uri.path}`
      this.#logger.error(message)
      throw new Error(message)
    }

    this.emit('fileStat', snippet.id)
    return Snippet.fromObject(snippet)
  }

  readDirectory (): [string, vscode.FileType][] {
    const state = this._context.globalState.get<State>(STATE_KEY)
    return state?.snippets.map((s) => [s.title, vscode.FileType.File]) || []
  }

  readFile (uri: vscode.Uri): Uint8Array {
    const snippet = this.stat(uri)
    return snippet.data
  }

  async writeFile (uri: vscode.Uri, content: Uint8Array) {
    /**
     * add new snippet
     */
    if (uri.path.slice(1) === 'New Clipboard Item') {
      const snippetName = await vscode.window.showInputBox({
        title: 'Clipboard Name',
        prompt: 'Please give your Clipboard Item a name (e.g. myReactHook.ts)'
      }).then((val) => val || 'New Clipboard Item')

      const snippet = new Snippet(
        this._workspaceId,
        path.basename(snippetName),
        content.toString()
      )

      /**
       * emit with delay so that VS Code can store the file and no prompt appears
       */
      setTimeout(() => this.emit('saveNewSnippet', snippet), 100)
      this.#logger.info(
        `New snippet add "${snippetName}"` + (this._workspaceId ? `, to workspace with id ${this._workspaceId}` : '')
      )
      return
    }

    const snippet = this.stat(uri)
    snippet.body = content.toString()
    this.#logger.info(`Updated snippet "${snippet.title}"`)
    setTimeout(() => this.emit('updateSnippet', snippet), 100)
  }

  // Not needed/implemented
  rename (): void {}
  delete (): void {}
  createDirectory (): void {}
}
