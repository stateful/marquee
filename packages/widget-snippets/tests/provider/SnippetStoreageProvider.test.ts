import path from 'path'
import SnippetStorageProvider from '../../src/provider/SnippetStorageProvider'

const enc = new TextEncoder()

describe('SnippetStorageProvider', () => {
  it('should return disposable for watch', () => {
    const provider = new SnippetStorageProvider({} as any)
    expect(typeof provider.watch().dispose).toBe('function')
  })

  it('stat: new Snippet', () => {
    const provider = new SnippetStorageProvider({} as any, 'foobar')
    const s = provider.stat({ path: '/New Clipboard Item'} as any)
    expect(s.workspaceId).toBe('foobar')
  })

  it('stat: throws if snippet can not be found', () => {
    const snippets = [{ path: '/foo/bar' }, { path: '/bar/foo' }]
    const context = { globalState: { get: jest.fn().mockReturnValue({ snippets }) } }
    const provider = new SnippetStorageProvider(context as any)

    expect.assertions(1)
    try {
      provider.stat({ path: '/foobar' } as any)
    } catch (err: any) {
      expect(err.message).toContain('Couldn\'t find snippet')
    }
  })

  it('should be able to also find old snippet types', () => {
    const snippets = [{ path: 'Untitled:1:0', id: '12345' }, { path: '/bar/foo' }]
    const context = { globalState: { get: jest.fn().mockReturnValue({ snippets }) } }
    const provider = new SnippetStorageProvider(context as any)
    const s = provider.stat({ path: '/12345/Untitled:1:0' } as any)
    expect(s.id).toBe('12345')
  })

  it('stat: should return file stat', () => {
    const snippets = [{ path: '/foo/bar' }, { path: '/bar/foo', id: 'foobar' }]
    const context = { globalState: { get: jest.fn().mockReturnValue({ snippets }) } }
    const provider = new SnippetStorageProvider(context as any)
    const s = provider.stat({ path: '/bar/foo' } as any)
    expect(s.id).toBe('foobar')
  })

  it('getDirectory', () => {
    const snippets = [{ title: '/foo/bar' }, { title: '/bar/foo', id: 'foobar' }]
    const context = { globalState: { get: jest.fn().mockReturnValue({ snippets }) } }
    const provider = new SnippetStorageProvider(context as any)
    expect(provider.readDirectory()).toMatchSnapshot()
  })

  it('readFile', () => {
    const snippets = [{ path: '/bar/foo', id: 'foobar', body: 'foobar' }]
    const context = { globalState: { get: jest.fn().mockReturnValue({ snippets }) } }
    const provider = new SnippetStorageProvider(context as any)
    expect(provider.readFile({ path: '/bar/foo' } as any)).toMatchSnapshot()
  })

  it('writeFile: new snippet', async () => {
    const provider = new SnippetStorageProvider({} as any)
    provider.emit = jest.fn()

    await provider.writeFile({ path: '/New Clipboard Item' } as any, enc.encode('some text'))
    await new Promise((resolve) => setTimeout(resolve, 100))

    expect(provider.emit).toBeCalledWith('saveNewSnippet', {
      archived: false,
      body: '115,111,109,101,32,116,101,120,116',
      createdAt: expect.any(Number),
      ctime: expect.any(Number),
      id: '39d2f858-3522-4053-beb8-b75a1defd1d2',
      mtime: expect.any(Number),
      origin: undefined,
      path: undefined,
      storagePath: path.sep + path.join('39d2f858-3522-4053-beb8-b75a1defd1d2', 'some input'),
      size: 34,
      title: 'some input',
      type: '1',
      branch: undefined,
      commit: undefined,
      gitUri: undefined,
      workspaceId: null,
    })
  })

  it('writeFile: update snippet', async () => {
    const snippets = [{ path: '/bar/foo', id: 'foobar', body: 'foobar', title: 'some title' }]
    const context = { globalState: { get: jest.fn().mockReturnValue({ snippets }) } }
    const provider = new SnippetStorageProvider(context as any)

    provider.emit = jest.fn()
    await provider.writeFile({ path: '/bar/foo' } as any, enc.encode('some text'))
    await new Promise((resolve) => setTimeout(resolve, 100))

    expect(provider.emit).toBeCalledWith('updateSnippet', {
      archived: false,
      body: '115,111,109,101,32,116,101,120,116',
      createdAt: expect.any(Number),
      ctime: expect.any(Number),
      id: 'foobar',
      mtime: expect.any(Number),
      origin: undefined,
      path: undefined,
      storagePath: path.sep + path.join('foobar', 'some title'),
      size: 6,
      title: 'some title',
      type: '1',
      branch: undefined,
      commit: undefined,
      gitUri: undefined,
      workspaceId: null,
    })
  })
})
