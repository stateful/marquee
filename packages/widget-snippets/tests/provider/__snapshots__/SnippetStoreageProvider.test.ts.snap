// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnippetStorageProvider getDirectory 1`] = `
Array [
  Array [
    "/foo/bar",
    "1",
  ],
  Array [
    "/bar/foo",
    "1",
  ],
]
`;

exports[`SnippetStorageProvider readFile 1`] = `
Uint8Array [
  102,
  111,
  111,
  98,
  97,
  114,
]
`;
